hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ExpiredContainerReplicaOpScrubber.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.hadoop.hdds.scm.container.replication;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hdds.HddsConfigKeys;
+import org.apache.hadoop.hdds.conf.ConfigurationSource;
+import org.apache.hadoop.hdds.scm.ScmConfigKeys;
+import org.apache.hadoop.hdds.scm.ha.SCMContext;
+import org.apache.hadoop.hdds.scm.ha.SCMService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Clock;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Background service to clean up expired container replica operations.
+ */
+public class ExpiredContainerReplicaOpScrubber implements SCMService {
```

Comment:
I have seen a few of these "service" classes, and there is a lot of repeated code in them, some of it tricky concurrent code with `wait()` and `notify()` etc. I think this is an opportunity to create an abstract class that that implements the service interfaces and provides a common set of code that other services can use going forward. If we create that, perhaps we can make the new replicationManager use it, and there was another service added recently for EC - BackGroundPipelineScrubber - its a very simple class like this one. I'm sure we can create a common "service" class these classes can extend.

Reply:
yea , i noticed this too. i wanted to create a separate jira to do this work.  but we can also do it in this jira, will try to do it

Reply:
Lets try to create the abstract class in this Jira and let this service use it. Also try to design it so that BackGroundPipelineScrubber and hopefully ReplicaitonManager could use it, but don't change them in this Jira. We can have a separate jira to refactor them to use it.